### ISSUE TITLE
请问能给出训练之后，单文本预测代码文件吗

### ISSUE THREAD

**Issue by vencentDebug** [thuiar/TEXTOIR#61]
加载训练之后的bin文件，如何推理是未知意图还是领域内

**Comment by vencentDebug** [comment]
自己写了一个不知道，对不对
from configs.base import ParamManager
from dataloaders.base import DataManager
from backbones.base import ModelManager
from methods import method_map
from utils.functions import save_results
import logging
import sys
import os
import datetime
import argparse

def parse_arguments():
    parser = argparse.ArgumentParser()

    # Hardcoded arguments based on the command line input
    parser.set_defaults(
        dataset='banking2',
        type='open_intent_detection',
        logger_name='Detection',
        log_dir='logs',
        known_cls_ratio=0.5,
        labeled_ratio=1.0,
        method='DA-ADB',
        train=False,
        pretrain=False,
        save_model=True,
        backbone='bert',
        config_file_name='DA-ADB.py',
        seed=0,
        gpu_id='0',
        pipe_results_path='pipe_results',
        data_dir='/wjl/query_cls/OOD/TEXROIR/data',
        output_dir='/home/sharing/disk1/zhaoshaojie/baseline_test/TEXTOIR',
        model_dir='models',
        result_dir='results',
        results_file_name='results_DA-ADB.csv',
        save_results=True,
        load_pretrained_method=None,
        loss_fct='CrossEntropyLoss'
    )

    args = parser.parse_args([])  # Empty list to use set_defaults
    return args

def set_logger(args):
    if not os.path.exists(args.log_dir):
        os.makedirs(args.log_dir)
    
    time = datetime.datetime.now().strftime('%Y-%m-%d-%H-%M-%S')
    file_name = f"{args.method}_{args.dataset}_{args.backbone}_{args.known_cls_ratio}_{args.labeled_ratio}_{time}.log"
    
    logger = logging.getLogger(args.logger_name)
    logger.setLevel(logging.DEBUG)

    return logger

def run(args, data, model, logger):
    method_manager = method_map[args.method]
    method = method_manager(args, data, model, logger_name = args.logger_name)
    
    if args.train:
        logger.info('Training Begin...')
        method.train(args, data)
        logger.info('Training Finished...')

    logger.info('Testing begin...')
    outputs = method.test(args, data)
    logger.info('Testing finished...')

    if args.save_results:
        logger.info('Results saved in %s', str(os.path.join(args.result_dir, args.results_file_name)))
        save_results(args, outputs)

import torch
import os

def load_pretrained_model(model, pretrained_path):
    """
    加载预训练模型权重
    
    Args:
    - model: 初始化的模型实例
    - pretrained_path: 预训练模型的完整路径
    
    Returns:
    - 加载权重后的模型
    """
    # 检查路径是否存在
    if not os.path.exists(pretrained_path):
        raise FileNotFoundError(f"Pretrained model path not found: {pretrained_path}")
    
    # 加载模型权重
    try:
        # 如果使用GPU训练的模型，可能需要map_location='cpu'
        state_dict = torch.load(pretrained_path, map_location='cpu')
        
        # 处理可能的键前缀问题
        # 有时保存的模型权重可能有额外的前缀，需要处理
        new_state_dict = {}
        for k, v in state_dict.items():
            # 去除可能的'module.'前缀
            if k.startswith('module.'):
                new_state_dict[k[7:]] = v
            else:
                new_state_dict[k] = v
        
        # 加载权重
        model.load_state_dict(new_state_dict)
        print(f"Pretrained model loaded successfully from {pretrained_path}")
    except Exception as e:
        print(f"Error loading pretrained model: {e}")
    
    return model
    
from transformers import BertTokenizer
import torch

def process_single_text(text, tokenizer, max_length=128):
    """
    将单个文本转换为BERT模型的输入格式
    
    Args:
    - text: 输入文本
    - tokenizer: BERT分词器
    - max_length: 最大序列长度
    
    Returns:
    - 模型输入字典
    """
    # 对文本进行编码
    encoding = tokenizer(
        text, 
        return_tensors='pt',  # 返回PyTorch张量
        max_length=max_length,  # 最大长度
        padding='max_length',   # 填充到最大长度
        truncation=True         # 超过最大长度则截断
    )
    
    return {
        'input_ids': encoding['input_ids'],
        'attention_mask': encoding['attention_mask'],
        'token_type_ids': encoding.get('token_type_ids', None)
    }

# def main():
sys.path.append('.')
args = parse_arguments()
logger = set_logger(args)

# BERT模型和分词器路径
bert_model = "/mnt/workspace/.cache/modelscope/hub/dienstag/chinese-bert-wwm-ext"
tokenizer = BertTokenizer.from_pretrained(bert_model, do_lower_case=True)

# 初始化参数和模型
logger.info('Open Intent Detection Begin...')
logger.info('Parameters Initialization...')
param = ParamManager(args)
args = param.args

logger.info('Data and Model Preparation...')
data = DataManager(args, logger_name = args.logger_name)
model_manager = ModelManager(args, data, logger_name = args.logger_name)
model = model_manager.set_model(args, pattern='bert')
optimizer, scheduler = model_manager.set_optimizer(model, data.dataloader.num_train_examples, args.train_batch_size, \
        args.num_train_epochs, args.lr, args.warmup_proportion)

# 将模型设置为评估模式
model.eval()

# 预处理文本
text = "你是机器人。转人工"
inputs = process_single_text(text, tokenizer)

# 获取模型预测
# with torch.no_grad():
#     # 确保inputs中的张量在与模型相同的设备上
#     inputs = {k: v.to(model.device) for k, v in inputs.items() if v is not None}
#     outputs = model(**inputs)
#     print(outputs)
#     print(outputs.shape)
    
    
#     # 根据你的模型结构，可能需要处理outputs
#     # 例如：获取分类结果
#     predictions = torch.argmax(outputs.logits, dim=1)
#     print("预测结果:", predictions)
with torch.no_grad():
    inputs = {k: v.to(model.device) for k, v in inputs.items() if v is not None}
    outputs = model(**inputs)
    print(outputs)
    
    # Iterate through the tuple and print shapes
    for i, output in enumerate(outputs):
        print(f"Output {i} type: {type(output)}")
        if isinstance(output, torch.Tensor):
            print(f"Output {i} shape: {output.shape}")

# if __name__ == '__main__':
#     main()
        


method_manager2 = method_map[args.method]
method = method_manager2(args, data, model_manager, logger_name = args.logger_name)
preds = method.open_classify(data, outputs[0])
preds
tensor([16], device='cuda:0')

features2 = outputs[1]
features2
tensor([[ 0.0051, -0.0524, -0.0407, -0.0768, -0.1367, -0.0060, -0.0449, -0.0732,
         -0.1105,  0.1584, -0.0625,  0.0172, -0.0605, -0.0042,  0.2171,  0.0437]],
       device='cuda:0')
       
   probs2, preds2 = F.softmax(features2.detach(), dim = 1).max(dim = 1)
   
   (tensor([0.0784], device='cuda:0'), tensor([14], device='cuda:0'))
